### ISSUE TITLE
Feature request: DllPlugin is not compatible with tree shaking

### ISSUE THREAD

**Issue by asapach** [webpack/webpack#7333]
## Feature request

**What is the expected behavior?**
Bundles built using `DllPlugin` should take advantage of tree shaking: module concatenation (scope hoisting) and dead code removal.

**What is motivation or use case for adding/changing the behavior?**
Currently `ModuleConcatenationPlugin` bails out on any optimizations when `DllPlugin` is being used, because all the modules are being flagged as "used" even if they are not entry points:
> `ModuleConcatenation bailout: Module is referenced by: DllPlugin`

Here's a trivial example:
```js
/** @file src/dep.js */

export function foo() { // this export is never used and should be removed
    console.log('foo');
}

export const bar = 'bar';
```

```js
/** @file src/index.js */

import {bar} from './dep'; // only `bar` export is being used

console.log(bar);
```

When webpack runs in production mode the modules are correctly concatenated and the unused export is successfully removed. But when `DllPlugin` is added to the list, none of the optimizations are taking place.

**How should this be implemented in your opinion?**
There should be a way to indicate in `DllPlugin` options which modules (exports) are actually being exposed in the public contract/manifest (i.e. "used"), so that `ModuleConcatenationPlugin` can do its work. Alternatively there could be a new plugin that would introduce the new semantics and the existing `DllPlugin` would maintain the existing behavior.

**Are you willing to work on this yourself?**
maybe

**Comment by sokra** [comment]
Yep send a PR adding an option to the plugin which only exposes modules mentioned in the entrypoint

**Comment by asapach** [comment]
Done: #7344

**Comment by asapach** [comment]
Merged in 8af0320c3ea18cabc4ca62585fc98815f128f09d
